Step4 - 계산기 UI

ghojeong opened the pull request:
계산기 UI 를 TDD 로 구현을 해보았습니다.

Toast 가 띄워졌는지 여부를 어떻게 테스트 코드로 작성해야할 지 막혀서 도움을 요청합니다.

ghojeong committed:
refactor: Operation 의 constructor 가 private 이 되지 않도록 수정

ghojeong committed:
feat: 숫자를 입력하는 기능 구현

ghojeong committed:
feat: 연산자를 입력하는 기능 구현

- 입력된 피연사자가 없을 때, 피연산자를 누르면 화면에 숫자가 보여야 한다.
- 입력된 피연산자가 없을 때, 연산자를 누르면 아무런 변화가 없어야 한다.
- 입력된 피연산자가 있을 때, 연산자를 누르면 기호가 화면에 보여야 한다.

ghojeong committed:
feat: 지우기 버튼 구현

- 입력된 수식이 없을 때, 지우기 버튼을 누르면 화면에 아무런 변화가 없어야 한다.
- 입력된 수식이 있을 때, 지우기 버튼을 누르면 마지막 입력이 지워져야 한다.

ghojeong committed:
feat: = 버튼 구현

- 입력된 수식이 완전할 때, 사용자가 = 버튼을 누르면 입력된 수식의 결과가 화면에 보여야 한다.
- 입력된 수식이 완전하지 않을 때, 사용자가 = 버튼을 누르면 토스트 메세지가 화면에 보여야 한다.

ghojeong commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,114 +1,331 @@
+        }
+    }
+
+    private fun assertToastDisplayed(message: String) {
+        onView(withText(message)).check(matches(isDisplayed()))
+    }
```
이렇게 하면, 해당하는 ToastMessage 가 띄워졌는지 확인할 수 있을 거라고 생각했는데,
Test 가 현재 Fail 하고 있습니다.
혼자서는 해결하기 어려워서 도움을 주시면 감사하겠습니다.

BeokBeok commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,114 +1,331 @@
+
+        // then: 화면에 89가 보여야 한다.
+        assertTextEquals("5 + 1")
+    }
```
주석이 위의 inputNumber99에서 사용된 것과 동일하네요.
주석을 테스트 코드에 맞게 변경해주세요.

ghojeong replied:
8992c8b 커밋에서 주석 변경하였습니다.
주석으로 설명을 하려니깐 코드와 주석이 불일치하는 경우가 생겨서 불편하네요.
클린 코드 책에서도 주석에 관하여 동일한 문제를 지적했던것 같습니다.

나중에라도 협업하다가 주석에 대한 수정을 깜빡하는 일이 생길까 두렵네요.
BeokBeok 님은 현업에서 주석을 어떻게 관리하시나요?

BeokBeok replied:
개발자마다 다르겠지만, 저도 클린 코드 책에서 적혀있듯이 주석을 좋아하지 않습니다.
물론 주석에 대해서 유지보수를 잘 하면 좋지만, 개인적으로 대부분 주석이 유지보수가 잘 되는 경우를 본 적이 없었습니다.
특히 다른 분들이 작성한 주석은 거의 건드리지 않기에, 시간이 지나면 지날수록 주석은 아무도 못 건드리는 영역이 되더라구요.

해서, 저는 주석을 작성하지도 TODO나 FIXME도 사용하지 않으며, 가급적이면 코드로 의미를 표현하려고 노력을 많이 합니다.

BeokBeok commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,114 +1,331 @@
+        // when: 지우기 버튼을 누르면
+        // then: 마지막 입력이 지워진다.
+        onView(withId(R.id.buttonDelete)).perform(click())
+        assertTextEquals("32 + ")
```
```suggestion
        // when: 지우기 버튼을 누르면
        onView(withId(R.id.buttonDelete)).perform(click())
        
        // then: 마지막 입력이 지워진다.
        assertTextEquals("32 + ")
```
이렇게 되어야할 것 같아요

ghojeong replied:
8992c8b 커밋에서 주석 변경하였습니다.

BeokBeok commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,114 +1,331 @@
+        onView(withId(R.id.textView)).check(matches(withText(expected)))
+    }
+
+    // NOTE: 참고한 링크 (https://stackoverflow.com/questions/32846738/android-testing-how-to-change-text-of-a-textview-using-espresso)
+    private fun replaceText(value: String): ViewAction = object : ViewAction {
+        override fun getConstraints(): Matcher<View> = CoreMatchers.allOf(
+            isDisplayed(),
+            isAssignableFrom(TextView::class.java)
+        )
+
+        override fun getDescription(): String {
+            return "replace text"
+        }
+
+        override fun perform(uiController: UiController, view: View) {
+            (view as TextView).text = value
+        }
+    }
```
replaceText 함수 활용 👍

BeokBeok commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -13,17 +14,71 @@ class MainActivity : AppCompatActivity() {
-        binding.button7.setOnClickListener { binding.textView.text = "7" }
-        binding.button8.setOnClickListener { binding.textView.text = "8" }
-        binding.button9.setOnClickListener { binding.textView.text = "9" }
+        setupNumberButtonListener()
+        setupArithmeticButtonListener()
+        setupFunctionButtonListener()
```
기능별 함수 분리 👍

BeokBeok commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -13,17 +14,71 @@ class MainActivity : AppCompatActivity() {
+            if (text.isBlank()) return@setOnClickListener
+            textView.text = text.dropLast(
+                text.last().toString()
+                    .toIntOrNull()?.let { 1 } ?: 3
```
1과 3이라는 숫자는 어떤 의미일까요?

ghojeong replied:
9f617da 커밋에서 상수명을 명시함으로써 의미를 알수  있도록 하였습니다.

BeokBeok commented on domain/src/main/java/edu/nextstep/camp/calculator/Operation.kt:
```diff
@@ -1,14 +1,9 @@
 package edu.nextstep.camp.calculator
 
-data class Operation private constructor(
+data class Operation constructor(
```
```suggestion
data class Operation(
```
constructor를 생략해도 될 것 같아요

ghojeong replied:
실수군요. 피드백 감사합니다.
8992c8b 커밋에서 반영하였습니다.

BeokBeok commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -13,17 +14,71 @@ class MainActivity : AppCompatActivity() {
+        buttonEquals.setOnClickListener {
+            if (textView.text.isBlank()) return@setOnClickListener
+            runCatching {
+                textView.text = calculate(textView.text.toString()).toString()
```
```suggestion
                textView.text = Calculator.calculate(textView.text.toString()).toString()
```
지금처럼 import를 이용하면 함수만 호출할 수 있으나, 이 함수를 직접 타고가지 않는 이상 MainActivity 내부에 구현된 함수로 오해하기가 쉽습니다.
명시적으로 Calculator를 통해서 함수를 호출하도록 변경해주시는 것이 좋습니다.

ghojeong replied:
8992c8b 커밋에서 반영하였습니다.
무조건 static import 를 한다고 깔끔한게 아니군요.
가독성에 관한 관점을 제시해주셔서 감사합니다.

BeokBeok commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -13,17 +14,71 @@ class MainActivity : AppCompatActivity() {
+        button6.setOnClickListener { textView.text = "${textView.text}6" }
+        button7.setOnClickListener { textView.text = "${textView.text}7" }
+        button8.setOnClickListener { textView.text = "${textView.text}8" }
+        button9.setOnClickListener { textView.text = "${textView.text}9" }
```
문자열을 조합하기 보다는 string 리소스를 활용해보면 어떨까요?
문구는 가급적이면 string 리소스를 활용해주시는 것이 좋습니다.

ghojeong replied:
👍 
[String Resources](https://developer.android.com/guide/topics/resources/string-resource) 잘 활용해야하는데, 깜빡했군요. 피드백주셔서 감사합니다.
d14fce7 커밋에서 반영하였습니다.

ghojeong committed:
test: 테스트를 위한 ToastMatcher 클래스 생성

태현님이 조언 주신 내용을 적용했습니다.
https://stackoverflow.com/a/33387980/6686126

ghojeong committed:
test: 잘못된 private 접근 제어자 제거

ghojeong commented on app/src/androidTest/java/edu/nextstep/camp/calculator/MainActivityTest.kt:
```diff
@@ -1,114 +1,332 @@
+        assertTextEquals("5")
+    }
+
+    // 입력된 수식이 불완전할 때, = 버튼을 누르면, 토스트 메세지가 화면에 보인다.
+    @Test
+    fun incompleteInput_pressEqualsButton() {
+        // given: 3 + 가 입력되었다.
+        onView(withId(R.id.textView)).perform(replaceText("3 + "))
+
+        // when: = 버튼을 누르면
+        onView(withId(R.id.buttonEquals)).perform(click())
+
+        // then: 토스트 메세지가 보여진다.
+        assertToastDisplayed(TOAST_MESSAGE)
+    }
+
+    // 0 으로 나누는 수식을 입력하고, = 버튼을 누르면, 토스트 메세지가 화면에 보인다.
+    @Test
+    fun divideBy0() {
+        // given: 2 / 0 가 입력되었다.
+        onView(withId(R.id.textView)).perform(replaceText("2 / 0"))
+
+        // when: = 버튼을 누르면
+        onView(withId(R.id.buttonEquals)).perform(click())
+
+        // then: 토스트 메세지가 보여진다.
+        assertToastDisplayed(TOAST_MESSAGE)
+    }
```
[조언 받은 링크](https://stackoverflow.com/a/33387980/6686126)대로 똑같이 따라해보았는데, 토스트 메세지 관련된 테스트가 통과되고 있지 못하고 있네요.
무언가 제가 잘못을 한것 같은데 잘 모르겠습니다. 도움을 요청합니다.

이 2 테스트 외에는 모든 테스트가 잘 통과합니다.

BeokBeok replied:
혹시 안드로이드 몇 버전으로 테스트 하셨나요?
아쉽지만 토스트 관련 UI테스트는 안드로이드 11 이상 및 target sdk 30 이상 환경에서는 오류가 있는 것 같네요 😭 

ref. `https://github.com/android/android-test/issues/803`

BeokBeok reviewed (changes requested):
이전 단계 피드백과 계산기 UI 미션하시느라 고생하셨습니다 👍 
고민해볼만한 의견들을 코맨트로 작성하였으니, 충분히 고민해보시고 도전해보세요 💪

ghojeong committed:
fix: 사소한 변경사항 적용

- 주석 변경
- 명시적으로 Calculator import; calcuate -> Calculator.calculate
- Operator 의 불필요한 constructor 키워드 제거

ghojeong committed:
refactor: string 리소스를 활용하여 하드코딩된 문자 제거

ghojeong committed:
refactor: 변수명을 통해 매직넘버(1, 3)에 의미 부여

ghojeong commented:
> 아쉽지만 토스트 관련 UI테스트는 안드로이드 11 이상 및 target sdk 30 이상 환경에서는 오류가 있는 것 같네요

안드로이드 버전을 낮추니깐 UI 테스트를 잘 통과하네요.
OneUI 4.0 을 적용한 갤럭시 S10 최신 업데이트가 안드로이드 12 인데 어떡하죠 😭 
토스트 관련 UI 테스트는 현업에서 적용하기 무척 어렵겠네요.

@BeokBeok  님은 현업에서, 안드로이드 버전별로 테스트를 통과하지 못하는 일이 생길때 어떻게 대처하시나요?
모든 버전의 테스트를 통과하지 못하더라도, 특정 버전에서 의도한대로 통과한다면 무시하고 진행하는 편인가요?

꼭 토스트가 아니더라도 안드로이트 최신 업데이트가 나올때마다 이런일이 종종 생길것 같아서 여쭈어봅니다.

BeokBeok commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -13,17 +13,73 @@ class MainActivity : AppCompatActivity() {
+    }
+
+    private fun setupNumberButtonListener() = with(binding) {
+        button0.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_0)}" }
+        button1.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_1)}" }
+        button2.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_2)}" }
+        button3.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_3)}" }
+        button4.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_4)}" }
+        button5.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_5)}" }
+        button6.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_6)}" }
+        button7.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_7)}" }
+        button8.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_8)}" }
+        button9.setOnClickListener { textView.text = "${textView.text}${getString(R.string.calculator_9)}" }
```
string resource 활용 👍 
string resource 중에서 format 기능을 활용해주시면, 문구에 대해서 좀 더 유연하게 처리할 수 있을 것 같아요
`%s%s` 와 같은 포멧으로 하나는 TextView의 text를 하나는 입력된 숫자값을 넣어주시면 될 것 같아요

BeokBeok commented on app/src/main/java/edu/nextstep/camp/calculator/MainActivity.kt:
```diff
@@ -13,17 +13,73 @@ class MainActivity : AppCompatActivity() {
+            runCatching {
+                textView.text = Calculator.calculate(textView.text.toString()).toString()
+            }.getOrElse {
+                Toast.makeText(this@MainActivity, getString(R.string.calculator_incomplete_expression), Toast.LENGTH_SHORT)
```
가로로 길어지는 경우에는 파라미터 단위로 개행해주시면 코드가 한 눈에 들어옵니다

BeokBeok reviewed (approved):
피드백을 잘 반영해주셨네요 👍 
정완님과 미션을 함께 진행하면서 저 또한 성장하게 된 것 같습니다 💪 
좋은 개발자가 되기 위한 소중한 밑거름이 되었길 바랍니다.

BeokBeok commented:
> @BeokBeok 님은 현업에서, 안드로이드 버전별로 테스트를 통과하지 못하는 일이 생길때 어떻게 대처하시나요?
모든 버전의 테스트를 통과하지 못하더라도, 특정 버전에서 의도한대로 통과한다면 무시하고 진행하는 편인가요?
꼭 토스트가 아니더라도 안드로이트 최신 업데이트가 나올때마다 이런일이 종종 생길것 같아서 여쭈어봅니다.

특히 UI 테스트가 버전을 많이 타는 것 같습니다. 😭 
질문 주신 부분에 대해서 답변을 드릴 수 없어서 죄송하지만, 가급적이면 UI 테스트는 거의 작성하지 않고 단위 테스트 위주로 작성하는 편입니다 😅 

UI 테스트 코드를 기피하게 된 이유는 아래와 같습니다.
- 자주 변경되는 UI에 따른 잦은 테스트 코드 변경
- 오래 걸리는 빌드 시간

만약, 변경이 거의 없으며 중요한 기능을 담당하는 화면이 있다면 UI 테스트 코드를 작성하면 좋을 것 같아요.

ref. https://tv.kakao.com/channel/3693125/cliplink/414072285
